File: web/os-image/de/alpenmachine.zig
const std = @import("std");
const linux = std.os.linux;
const draw = @import("draw.zig");
const awidget = @import("awidget.zig");
const wire = @import("wire.zig");

fn nowSec() i64 {
    var ts: linux.timespec = undefined;
    _ = linux.clock_gettime(.REALTIME, &ts);
    return @intCast(ts.sec);
}

fn readFile(path: [*:0]const u8, buf: []u8) []const u8 {
    const fd_raw = linux.open(path, .{ .ACCMODE = .RDONLY }, 0);
    if (linux.errno(fd_raw) != .SUCCESS) return &.{};
    const fd: i32 = @intCast(fd_raw);
    const n = linux.read(fd, buf.ptr, buf.len);
    _ = linux.close(fd);
    if (linux.errno(n) != .SUCCESS) return &.{};
    return buf[0..n];
}

fn firstF64(data: []const u8) f64 {
    var i: usize = 0;
    while (i < data.len and (data[i] == ' ' or data[i] == '\n' or data[i] == '\t')) i += 1;
    const start = i;
    while (i < data.len and data[i] != ' ' and data[i] != '\n' and data[i] != '\t') i += 1;
    return std.fmt.parseFloat(f64, data[start..i]) catch 0;
}

pub fn main() void {
    var c: awidget.AwClient = undefined;
    if (!awidget.open(&c, "machine", 250, 92, -28, 26, 0)) linux.exit(1);

    var rbuf: [4096]u8 = undefined;
    var last: i64 = 0;
    while (true) {
        var in: wire.AwMsg = undefined;
        if (awidget.poll(&c, &in, 200) < 0) linux.exit(0);
        const now = nowSec();
        if (now == last) continue;
        last = now;

        const up = firstF64(readFile("/proc/uptime", &rbuf));
        const load = firstF64(readFile("/proc/loadavg", &rbuf));

        var mt: i64 = 0;
        var ma: i64 = 0;
        const mem = readFile("/proc/meminfo", &rbuf);
        var it = std.mem.tokenizeAny(u8, mem, " \n");
        while (it.next()) |k| {
            const v_str = it.next() orelse break;
            const _kb = it.next() orelse break;
            _ = _kb;
            const v = std.fmt.parseInt(i64, v_str, 10) catch 0;
            if (std.mem.eql(u8, k, "MemTotal:")) mt = v;
            if (std.mem.eql(u8, k, "MemAvailable:")) ma = v;
        }

        draw.frame(&c.buf, "machine");
        var buf: [96]u8 = undefined;
        const s1 = std.fmt.bufPrint(&buf, "up {d: >3}s   load {d:.2}", .{ @as(u32, @intFromFloat(up)), load }) catch unreachable;
        _ = draw.text(&c.buf, 12, 26, s1, 0xd8dbe2, 1);
        const s2 = std.fmt.bufPrint(&buf, "mem {d}/{d}M", .{ @divTrunc(mt - ma, 1024), @divTrunc(mt, 1024) }) catch unreachable;
        _ = draw.text(&c.buf, 12, 44, s2, 0xd8dbe2, 1);
        _ = draw.text(&c.buf, 12, 62, "linux 7.1.3 i686 real", 0x8890a0, 1);
        awidget.commit(&c);
    }
}

File: web/os-image/de/reference/alpenmachine.c
/*
 * alpenmachine — uptime, load and memory, read from the real /proc.
 *
 * A separate process reading a separate kernel interface, which is the
 * whole point: nothing here is the desktop's business.
 */
#include <stdio.h>
#include <string.h>
#include <time.h>

#include "awidget.h"

int main(void) {
  AwClient c;
  if (aw_open(&c, "machine", 250, 92, -28, 26, 0) < 0) return 1;

  time_t last = 0;
  for (;;) {
    AwMsg in;
    if (aw_poll(&c, &in, 200) < 0) return 0;
    time_t now = time(0);
    if (now == last) continue;
    last = now;

    double up = 0, load = 0;
    long mt = 0, ma = 0;
    FILE *f = fopen("/proc/uptime", "r");
    if (f) { if (fscanf(f, "%lf", &up) != 1) up = 0; fclose(f); }
    f = fopen("/proc/meminfo", "r");
    if (f) {
      char k[64];
      long v;
      while (fscanf(f, "%63s %ld kB\n", k, &v) == 2) {
        if (!strcmp(k, "MemTotal:")) mt = v;
        if (!strcmp(k, "MemAvailable:")) ma = v;
      }
      fclose(f);
    }
    f = fopen("/proc/loadavg", "r");
    if (f) { if (fscanf(f, "%lf", &load) != 1) load = 0; fclose(f); }

    aw_frame(&c.buf, "machine");
    char buf[96];
    snprintf(buf, sizeof buf, "up %3ds   load %.2f", (int)up, load);
    aw_text(&c.buf, 12, 26, buf, 0xd8dbe2, 1);
    snprintf(buf, sizeof buf, "mem %ld/%ldM", (mt - ma) / 1024, mt / 1024);
    aw_text(&c.buf, 12, 44, buf, 0xd8dbe2, 1);
    aw_text(&c.buf, 12, 62, "linux 7.1.3 i686 real", 0x8890a0, 1);
    aw_commit(&c);
  }
}
